src/wallet/rpcdump.cpp

```diff
@@ -549,6 +549,57 @@ UniValue dumpwallet_impl(const UniValue& params, bool fHelp, bool fDumpZKeys)
     return exportfilepath.string();
 }
 
+class AddSpendingKeyToWallet : public boost::static_visitor<bool>
+{
+private:
+    CWallet *m_wallet;
+public: 
+    AddSpendingKeyToWallet(CWallet *wallet) : m_wallet(wallet) {}
+
+    bool operator()(const libzcash::SproutSpendingKey &sk) const {
+        auto addr = sk.address();
+        // Don't throw error in case a key is already there
+        if (m_wallet->HaveSpendingKey(addr)) {
+            return true;
+        } else {
+            m_wallet->MarkDirty();
+
+            if (!m_wallet-> AddZKey(sk)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
+            }
+
+            m_wallet->mapZKeyMetadata[addr].nCreateTime = 1;
+            
+            return false;
+        }
+    }
+    
+    bool operator()(const libzcash::SaplingSpendingKey &sk) const {
+        auto fvk = sk.full_viewing_key();
+        auto addr = sk.default_address();
+        {
+            // Don't throw error in case a key is already there
+            if (m_wallet->HaveSaplingSpendingKey(fvk)) {
+                return true;
+            } else {
+                m_wallet->MarkDirty();
+    
+                if (!m_wallet-> AddSaplingZKey(sk)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
+                }
+    
+                m_wallet->mapSaplingZKeyMetadata[addr].nCreateTime = 1;
```

Comment:
Hmm, we should think about how this metadata is used, and whether it should index on `addr` or `fvk`.

Reply:
Will it still be used on addresses? Not a lot of instances in the code

Reply:
Okay, this data is stored in order to figure out when rescanning how many blocks can be skipped (that won't contain transactions for any of the known addresses). So I think it makes sense to store `SaplingFullViewingKey` instead of `SaplingPaymentAddress`, as `fvk` uniquely corresponds to the key that will be used to scan (both for incoming transactions, and in future for recovering outgoing transaction data). However, this is entirely an in-memory decision, so we can just make an issue, and handle it later (when we update `ScanForWalletTransactions()` to catch all Sapling (and Sprout) transactions without needing to reindex).